virttest/graphical_console.py

```diff
@@ -0,0 +1,97 @@
+"""
+Graphical Console
+"""
+
+import time
+
+
+def uniform_linear(duration, rate):
+
+    def motion(src, dst):
+        pass
+
+    return motion
+
+
+class DummyConsole(object):
+
+    """
+    Dummy console
+    """
+
+    KEY_DOWN = True
+    KEY_UP = False
+
+    BTN_DOWN = True
+    BTN_UP = False
+
+    SCROLL_FORWARD = 1
+    SCROLL_BACKWARD = -1
+
+    def __init__(self, vm, logfile=None):
+        self._vm = vm
+
+    def key_down(self, key):
+        self._key_event(self.KEY_DOWN, key)
+
+    def key_up(self, key):
+        self._key_event(self.KEY_UP, key)
+
+    def key_press(self, key, hold_time=0):
+        if isinstance(key, str):
+            key = [key]
+        self.key_down(key)
+        if hold_time:
+            time.sleep(hold_time)
+        self.key_up(key)
+
+    def mouse_down(self, btn):
+        self._btn_event(self.BTN_DOWN, btn)
+
+    def mouse_up(self, btn):
+        self._btn_event(self.BTN_UP, btn)
+
+    def mouse_press(self, btn, hold_time=0):
+        self.mouse_down(btn)
+        if hold_time:
+            time.sleep(hold_time)
+        self.mouse_up(btn)
+
+    def mouse_move(self, pos, motion=None):
+        for _pos in motion(pos):
+            self._motion_event(_pos)
+
+    def _key_event(self, event, key):
+        raise NotImplementedError()
+
+    def _btn_event(self, event, btn):
+        raise NotImplementedError()
+
+    def _scroll_event(self, event):
+        raise NotImplementedError()
+
+    def _motion_event(self, pos):
+        raise NotImplementedError()
+
+    def close(self):
+        pass
+
+
+class QMPInputConsole(DummyConsole):
+
+    """
+    QMP Input console.
+    """
+
+    def _key_event(self, down, key):
+        self._vm.monitor.input_send_key_event(down, key)
```

Comment:
Would you help indicate where are `input_send_key_event` and `input_send_btn_event` defined?

Reply:
sorry for that, I missed one patch for there two functions, will add them now~